compiler/rustc_typeck/src/astconv/mod.rs

```diff
@@ -1796,7 +1801,55 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
-                        self.prohibit_generics(slice::from_ref(assoc_segment).iter());
+                        self.prohibit_generics(slice::from_ref(assoc_segment).iter(), |err| {
+                            err.note("enum variants can't have type parameters");
+                            let type_name = tcx.opt_item_name(adt_def.did());
+                            let the_enum = type_name.map(|n| format!("enum `{n}`")).unwrap_or_else(|| "the enum".to_string());
+                            let msg = format!("you might have meant to specity type parameters on {the_enum}");
+                            let Some(args) = assoc_segment.args else { return; };
+                            let args_span = assoc_segment.ident.span.shrink_to_hi().to(args.span_ext);
+                            let Ok(snippet) = tcx.sess.source_map().span_to_snippet(args_span) else {
+                                err.note(&msg);
+                                return;
+                            };
+                            let (qself_sugg_span, is_self) = if let hir::TyKind::Path(hir::QPath::Resolved(_, ref path)) = qself.kind {
+                                // If the path segment already has type params, we want to overwrite
+                                // them.
+                                match &path.segments[..] {
+                                    [.., segment, _] => (
+                                        segment.ident.span.shrink_to_hi().to(segment.args.map_or(
+                                            segment.ident.span.shrink_to_hi(),
+                                            |a| a.span_ext)),
+                                        false,
+                                    ),
+                                    [segment] => (
+                                        segment.ident.span.shrink_to_hi().to(segment.args.map_or(
+                                            segment.ident.span.shrink_to_hi(),
+                                            |a| a.span_ext)),
+                                        kw::SelfUpper == segment.ident.name,
+                                    ),
+                                    _ => unreachable!(),
+                                }
+                            } else {
+                                err.note(&msg);
+                                return;
+                            };
+                            let Some(type_name) = type_name else {
+                                err.note(&msg);
+                                return;
+                            };
+                            let suggestion = vec![
+                                if is_self {
+                                    // Account for people writing `Self::Variant::<Args>`, where
+                                    // `Self` is the enum.
+                                    (qself.span, format!("{type_name}{snippet}"))
```

Comment:
Should we have the same `TODO` here about checking that `type_name` accepts generics?